torchtune/modules/kv_cache.py

```diff
@@ -91,18 +97,14 @@ def update(
                 f", but found new key tensors with batch size {k_val.shape[0]}!"
             )
 
-        if (self.size + seq_len) > self.k_cache.shape[2]:
-            raise ValueError(
-                f"The current cache has been setup with a sequence length of {self.k_cache.shape[2]}"
-                f", but the cache has reached a sequence length of {(self.size + seq_len)}!"
-            )
-        cache_pos = torch.arange(self.size, self.size + seq_len, device=k_val.device)
-        self.size += seq_len
-
+        assert (self.cache_pos[0] + seq_len) <= self.k_cache.shape[2]
         k_out = self.k_cache
         v_out = self.v_cache
 
-        k_out.index_copy_(2, cache_pos, k_val)
-        v_out.index_copy_(2, cache_pos, v_val)
+        k_out.index_copy_(2, self.cache_pos[:seq_len], k_val)
+        v_out.index_copy_(2, self.cache_pos[:seq_len], v_val)
```

Comment:
Not a blocker for this PR but apparently index_copy_ isn't supported on MPS?

Reply:
every day we fall further from the grace of steve jobs

Reply:
Any objection to swapping with the non-inplace version?

Reply:
A steve jobs fan I see?

fine

Reply:
actually no
```bash
        k_out[:, :, self.cache_pos[:seq_len]] = k_val
>       v_out[:, :, self.cache_pos[:seq_len]] = v_out
E       RuntimeError: unsupported operation: some elements of the input tensor and the written-to tensor refer to a single memory location. Please clone() the tensor before performing the operation.
```